File: ruby/10_from_push_to_pull/lib/values_of_month.rb
class ValuesOfMonth
  attr_reader :date_of_month, :transactions_of_month, :preceding_balance
  private :date_of_month, :transactions_of_month, :preceding_balance

  def initialize(date_of_month = Date.today, transactions_of_month = [], preceding_balance = 0)
    @date_of_month = date_of_month
    @transactions_of_month = transactions_of_month
    @preceding_balance = preceding_balance
  end

  def balance
    transactions_of_month.map(&:amount).inject(preceding_balance, :+)
  end

  def average_balance
    transactions_of_month.inject(preceding_balance) { |average_balance, transaction|
      average_balance + rate_of(transaction)
    }.to_i
  end

  private
  def rate_of(transaction)
    days_of_month = date_of_month.day
    counting_days = days_of_month - transaction.date.day + 1
    rate = counting_days.to_f / days_of_month.to_f
    transaction.amount * rate
  end
end

File: csharp/10_FromPushToPull/ValuesOfMonth.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common;

namespace FromPushToPull
{
    public class ValuesOfMonth
    {
        private readonly DateTime dateOfMonth;
        private readonly IList<Transaction> transactionsOfMonth;
        private readonly int precedingBalance;

        public int Balance
        {
            get
            {
                int balance = precedingBalance;
                foreach (Transaction transaction in transactionsOfMonth)
                {
                    balance += transaction.Amount;
                }
                return balance;
            }
        }

        public int AverageBalance
        {
            get
            {
                double averageBalance = precedingBalance;
                foreach (Transaction transaction in transactionsOfMonth)
                {
                    averageBalance += RateOf(transaction);
                }
                return (int)averageBalance;
            }
        }

        public ValuesOfMonth() : this(DateTime.Now, new List<Transaction>(), 0) { }

        public ValuesOfMonth(DateTime dateOfMonth, IList<Transaction> transactionsOfMonth, int precedingBalance)
        {
            this.dateOfMonth = dateOfMonth;
            this.transactionsOfMonth = transactionsOfMonth;
            this.precedingBalance = precedingBalance;
        }

        private double RateOf(Transaction transaction)
        {
            int countingDays = (dateOfMonth - transaction.Date).Days + 1;
            double rate = (double)countingDays / dateOfMonth.Day;
            return transaction.Amount * rate;
        }
    }
}
